### PR TITLE
Implement image sending from sender

### PR DESCRIPTION
**Pull request by bitfriend**
I tested image sending.
I confirmed message history appears in chat.
But image is not displayed.
Because the feature that fetches the received image is not implemented on flutter & rust.
Now implementing it.

### CODE REVIEW

**Comment by bitfriend** [comment]
Reformatted by rust code

**Comment by bitfriend** [comment]
Reformatted by rust code

File: native/effektio/Cargo.toml
Comment (gnunicorn) [inline native/effektio/Cargo.toml:56]: I appreciate you finding out there is an open msc... but there's also an already implemented version we should be doing instead...

**Review by gnunicorn — CHANGES_REQUESTED** [review]
we don't want to send using the newly, yet to be approved MSC but instead use the existing "attachment" types.

File: native/effektio/Cargo.toml
Comment (bitfriend) [reply native/effektio/Cargo.toml:56]: Noted. Will check it.

File: native/effektio/src/api/room.rs
Comment (bitfriend) [reply native/effektio/src/api/room.rs:227]: Noted. Will check it.

File: native/effektio/Cargo.toml
Comment (bitfriend) [reply native/effektio/Cargo.toml]: Just pushed #0eb3936393570afddc6ba244aaa9ab4ba7809277 (Replace msc3552 with standard on image send)

File: native/effektio/src/api/room.rs
Comment (bitfriend) [reply native/effektio/src/api/room.rs]: Just pushed #0eb3936393570afddc6ba244aaa9ab4ba7809277 (Replace msc3552 with standard on image send)

File: native/effektio/src/api/room.rs
Comment (gnunicorn) [inline native/effektio/src/api/room.rs:204]: message isn't used anymore.

```suggestion
```

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:142]: `message` isn't used anymore.

**Review by gnunicorn — APPROVED** [review]
code looks fine, and seems to work, but the UI shows the now-outdated error message still:
Screenshot_1653421761

**Comment by bitfriend** [comment]
The `get_media_content` function of `matrix sdk` returns image binary not file URL.
So I couldn't display image with URL on flutter side.
I implemented custom image message builder using `Image.memory`.

The user credential is needed before encrypted image can be decrypted.
So I implemented `CURRENT_CLIENT` variable globally, it keeps active `matrix client` when user does login.

And on rust side, I implemented the class that keeps image-specific info, excluding standard message info.
It is `ImageDescription`.

I used `string_validator` library to check if image URL is valid on flutter side.
Because the image that user just sent is displayed from local not remote.

**Comment by bitfriend** [comment]
`RUNTIME.block_on` must be replaced with `RUNTIME.spawn`, before we can avoid deadlock when opening `ChatScreen` page.
I replaced it in `bin_data` of `ImageDescription`.
But I didn't replace in `image_description` of `RoomMessage`.
Now I can see bug that image is not displayed if above function replaced.
This is problem that should be fixed as soon.

File: native/effektio/src/api/client.rs
Comment (gnunicorn) [inline native/effektio/src/api/client.rs:50]: we can't do that, sorry. The client is hold by the surrounding SDK and the code may not assume that there is a global to store the current client. Why? Because the SDK is meant to allow for multi-client support (even if the UI doesn't implement it yet).

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:96]: in rust all fields are private, unless prefixed with `pub`. `_`-prefixes are only used for internal variables, that aren't actually needed. Let's stick to that model and stay with regular names instead.

```suggestion
    bin_data: Vec<u8>,
    name: String,
    mimetype: Option<String>,
    size: u64,
    width: Option<u64>,
    height: Option<u64>,
```

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:100]: this shouldn't copy here but only on request actually call to client and fetch the image and return it.

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:91]: We shouldn't hold this constantly in memory here.

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:61]: we shouldn't do this here but only upon request of the actual function call.

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:50]: rather than assuming a global, we should put `client` onto the room-message (and the image description), so we can do:

   ```suggestion
                let client = self.client.clone();
```

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:105]: can we have a better name for this function? How about `image_binary` or something?

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:115]: Please add doc comments to all new functions in this file.

**Review by gnunicorn — CHANGES_REQUESTED** [review]
Please revert from using a global client and instead pass a `client: self.client.clone()` when creating messages from the room.

Further comments inline.

File: native/effektio/src/api/messages.rs
Comment (bitfriend) [reply native/effektio/src/api/messages.rs:96]: Yes, rust recognize `_` as private property in struct.
By the way, `ImageDescription` must keep the getter functions about these fields.
So I appended the prefix `_` to properties.

File: native/effektio/src/api/messages.rs
Comment (bitfriend) [reply native/effektio/src/api/messages.rs:100]: Will move fetching function from message to client?

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:105]: Great idea

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [reply native/effektio/src/api/messages.rs:96]: Sure, `_`-prefixes are acceptable, but are unnecessary (and make it needlessly harder to read), so I'd prefer we dropped them. Of course that doesn't mean we don't need the getter functions, that's totally okay. Was just talking about style...

File: native/effektio/src/api/client.rs
Comment (bitfriend) [reply native/effektio/src/api/client.rs:50]: Yes, correct.
It doesn't look smart for me too.
But I guess on rust side there is no way to manage the multiple clients and set active to some client.
Alternatively on flutter side there is `currentClient` of `EffektioSdk`.
If we go and back between rust and flutter, we can get delay.
Before flutter can deliver current client to rust, flutter must pass token to rust and rust must login with it, so it takes a bit of time to perform this login, because login is operation via network.
Really I tested it and confirmed delay when opening `ChatScreen` page.
So I just wanted to finish all operations in rust, without go and back between rust and flutter.

Okay.
Will work from `currentClient` of `EffektioSdk` of flutter side.

File: native/effektio/src/api/messages.rs
Comment (bitfriend) [reply native/effektio/src/api/messages.rs:96]: Understood.
Will change property name and keep getter function name.

File: native/effektio/src/api/client.rs
Comment (gtalha07) [reply native/effektio/src/api/client.rs:50]: Did you tested it in '''release''' build?

Comment (bitfriend) [reply native/effektio/src/api/client.rs:50]: No, I tested in debug build

**Comment by bitfriend** [comment]
1. In case that `_insertMessage` is wrapped into `setState`.
What if the 1st image binary reaches before `initState` completes in `ChatScreen` page?

2. In case that `_insertMessage` is not wrapped into `setState`.
What if the fetching image binary reaches after `initState` completed in `ChatScreen` page?

Without mutex, it is impossible to solve these problems atomically.
So I added `mutex` to dependency and implemented mutex for state updating.

PS:
As you see, now this app is handling the image caching in low level directly.
So mutex and image custom builder was introduced.
If it is possible to use remote URL (that means image is not encrypted), we could remove all these complicated things.
But we are targeting the encrypted images, so we must handle the image caching directly.

**Comment by bitfriend** [comment]
Hi, Ben.
Your commit (https://github.com/effektio/effektio/pull/113/commits/211b4f53de77df15387ef5ade22e285b16c14562) to my branch is working well.
Thank you for great help again.

**Comment by bitfriend** [comment]
I used `path.join` to prepare file download path.
This function depends on `libandroidicu.so` and I included this library into `jniLibs`.
I could get this library for only `x86_64`, not `x86`/`arm64-v8a`/`armeabi-v7a`.

I used `fluttertoast` to show error in case that user clicked the downloaded file and there is no associated with that file.

I moved the image/file reading from message to room.
Because room on rust side keeps client and `ChatScreen.dart` contains room object.
So we can keep `messages.rs` simple as original (I had introduced matrix client to read image/file in `messages.rs`).

**Comment by gnunicorn** [comment]
> I used fluttertoast to show error in case that user clicked the downloaded file and there is no associated with that file.

we've just removed fluttertoast (see #119 ), please use the default Material Snackbar.

> I used path.join to prepare file download path on flutter side.
> This function depends on libandroidicu.so and I included this library into jniLibs.

Let's not. Please use rust's `path`-stuff for anything localised - just add a new method.

**Comment by bitfriend** [comment]
> > I used fluttertoast to show error in case that user clicked the downloaded file and there is no associated with that file.
> 
> we've just removed fluttertoast (see #119 ), please use the default Material Snackbar.
> 
> > I used path.join to prepare file download path on flutter side.
> > This function depends on libandroidicu.so and I included this library into jniLibs.
> 
> Let's not. Please use rust's `path`-stuff for anything localised - just add a new method.

Great idea.
Will do.

**Comment by bitfriend** [comment]
`ubuntu-latest` was upgraded from `20220529` to `20220605`.
Now we are getting the error [unable to find library -lgcc](https://github.com/bbqsrc/cargo-ndk/issues/22).
2 changes made to fix this bug.
1. Downgrade `cargo-ndk`. Replace `cargo install cargo-ndk` with `cargo install cargo-ndk --version 2.6.0` in `Makefile.toml` of root directory.
2. Add this step prior to `cargo make --profile release android-arm`
```
      - name: Copy ndk utilities for cargo make
        run: |
          rmdir -r /usr/local/lib/android/sdk/ndk-bundle/toolchains
          mkdir -p /usr/local/lib/android/sdk/ndk-bundle/toolchains
          cp -R ${{ steps.install-ndk-22.outputs.ndk-path }}/toolchains/llvm /usr/local/lib/android/sdk/ndk-bundle/toolchains
```

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:99]: we might want to consider switching this into an `Enum`-type... but later...

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:105]: What if there is no image/file? I assume this is then `None`? Shouldn't this be `Option` rather than `Result` then?

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:134]:
```suggestion
    /// file size in byte
```

File: native/effektio/api.rsh
Comment (gnunicorn) [inline native/effektio/api.rsh:184]: why is this function not on `image` or `RoomMessage`?

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [reply native/effektio/src/api/messages.rs:53]: But more interestingly... if `size` might be optional... why not just make it optional in the API, too ... same as width and height... size `0` is not the same as an unknown size, after all...

File: native/effektio/src/api/messages.rs
Comment (gnunicorn) [inline native/effektio/src/api/messages.rs:64]: same as above re if-let

File: native/effektio/src/api/room.rs
Comment (gnunicorn) [inline native/effektio/src/api/room.rs:344]: please don't `unwrap` in actual code (tests are fine). Instead always use `.expect(MSG)` with a message stating, why we can be sure this never panics. e.g. : 

```suggestion
                                .set_custom_value(&key, path.to_str().expect("Path was generated from strings. Must be string").as_bytes().to_vec())
```

File: native/effektio/src/api/room.rs
Comment (gnunicorn) [inline native/effektio/src/api/room.rs:380]: same as above. please don't unwrap. always `expect`.

File: native/effektio/src/api/room.rs
Comment (gnunicorn) [inline native/effektio/src/api/room.rs:388]: Pretty sure, these should all be `None` rather than `Ok` of an empty string.

**Review by gnunicorn — APPROVED** [review]
There's a bunch of minor API design and internal style remarks. Please fix at your own peril and merge.

File: native/effektio/api.rsh
Comment (gnunicorn) [reply native/effektio/api.rsh:99]: let's make that into an issue for later, once the PR is merged.

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:105]: In fact, I was planning to user these functions for only file/image.
Will fix.

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:117]: Yes, you are correct.

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:134]: Yes, you are correct

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:105]: I wanted to abstract image description and file description into a single struct.
But they have different fields, so I couldn't abstract them.

File: native/effektio/api.rsh
Comment (bitfriend) [reply native/effektio/api.rsh:184]: Flutter keeps room as member variable in UI.
Flutter doesn't keep room message as variable, because it may have to load so many messages in screen.
So I wanted to get image binary data from member variable.
Getting temporary variable for each message may be slow.

Comment (gnunicorn) [reply native/effektio/api.rsh:184]: yeah, fair enough then. keep it the way it is.

File: native/effektio/src/api/messages.rs
Comment (bitfriend) [reply native/effektio/src/api/messages.rs:45]: Cool

File: native/effektio/src/api/messages.rs
Comment (bitfriend) [reply native/effektio/src/api/messages.rs:64]: Cool

File: native/effektio/src/api/room.rs
Comment (bitfriend) [reply native/effektio/src/api/room.rs:344]: Will fix

**Pull request closed without merging**
